lessons-01 (#85)
mimuw-jnp2-rust/mimuw-jnp2-rust.github.io: content/lessons/01_introduction/index.md

- What does it mean that a language is "fast"? Is Python fast (tricky question)?
  Is Java fast? How to compare how "fast" is Rust next to C++?
- What does it mean that a language is "safe"? Is typing enough? What are non-trivial
  ways to make a bug, in C++ and in Python?
- Look into the examples of the source code of the Rust compiler:
  - [a few helper functions](https://github.com/rust-lang/rust/blob/master/compiler/rustc_borrowck/src/path_utils.rs) with short implementations,
  - the [docs](https://doc.rust-lang.org/stable/nightly-rustc/rustc_borrowck/index.html) of this crate,
  - some more complicated code: [match exhaustiveness](https://github.com/rust-lang/rust/blob/master/compiler/rustc_pattern_analysis/src/usefulness.rs),
  - [example PR](https://github.com/rust-lang/rust/issues/83104) that fixes an error suggestion.

## Why learn Rust?

Even if you don't end up using Rust, learning it expands your horizons:

- It helps especially with the awareness of what you can and can't do in concurrent applications.
- It helps you understand memory management and learn its good practices.
- the language is designed to avoid decades-long issues present in other languages.

>   It takes [inspiration](https://doc.rust-lang.org/reference/influences.html) from the best parts of other languages
>   to avoid the worst parts. Learning them shows what works.

Author: ```suggestion
  It takes [inspiration](https://doc.rust-lang.org/reference/influences.html) from the best parts of other languages
  to avoid the worst parts. Learning those parts show how to have better code design.
```
